arangod/Replication2/StateMachines/Document/DocumentFollowerState.cpp

```diff
@@ -375,220 +407,57 @@ auto DocumentFollowerState::handleSnapshotTransfer(
+    // the lock. Likewise, resign must shutdown the runtime inside the lock,
+    // _after_ it has set the resigning flag.
+    if (_resigning.load()) {
+      return Result{TRI_ERROR_REPLICATION_REPLICATED_LOG_FOLLOWER_RESIGNED};
+    }
 
-        return basics::catchToResultT([&]() -> std::optional<LogIndex> {
-          std::optional<LogIndex> releaseIndex;
+    futures::Promise<ResultT<std::optional<LogIndex>>> promise;
+    auto future = promise.getFuture();
+    _runtime->dispatch<actor::message::ApplyEntriesMessages>(
+        _applyEntriesActor, _applyEntriesActor,
+        actor::message::ApplyEntries{.entries = std::move(ptr),
+                                     .promise = std::move(promise)});
+
+    return std::move(future).thenValue(
+        [me =
+             weak_from_this()](ResultT<std::optional<LogIndex>> res) -> Result {
+          if (res.fail()) {
+            return res.result();
+          }
 
-          while (auto entry = ptr->next()) {
-            if (self->_resigning) {
-              // We have not officially resigned yet, but we are about to.
-              // So, we can just stop here.
-              break;
-            }
-            auto [index, doc] = *entry;
-
-            auto currentReleaseIndex = std::visit(
-                [&data, index = index](
-                    auto const& op) -> ResultT<std::optional<LogIndex>> {
-                  return data.applyEntry(op, index);
-                },
-                doc.getInnerOperation());
-
-            if (currentReleaseIndex.fail()) {
-              TRI_ASSERT(self->_errorHandler
-                             ->handleOpResult(doc.getInnerOperation(),
-                                              currentReleaseIndex.result())
-                             .fail())
-                  << currentReleaseIndex.result()
-                  << " should have been already handled for operation "
-                  << doc.getInnerOperation()
-                  << " during applyEntries of follower " << self->gid;
-              LOG_CTX("0aa2e", FATAL, self->loggerContext)
-                  << "failed to apply entry " << doc
-                  << " on follower: " << currentReleaseIndex.result();
-              TRI_ASSERT(false) << currentReleaseIndex.result();
-              FATAL_ERROR_EXIT();
-            }
-            if (currentReleaseIndex->has_value()) {
-              releaseIndex = std::move(currentReleaseIndex).get();
+          auto self = me.lock();
+          if (self == nullptr) {
+            return {TRI_ERROR_REPLICATION_REPLICATED_LOG_FOLLOWER_RESIGNED};
+          }
+          auto index = res.get();
+          if (index.has_value()) {
+            // The follower might have resigned, so we need to be careful when
+            // accessing the stream.
+            auto releaseRes = basics::catchVoidToResult([&] {
+              // TODO - is this getStream call actually safe?
```

Comment:
Yes, from that perspective it should be safe, but then why do we have this catchVoidToResult? This indicates to me, that this code was written with some assumption what might (no longer?) be true. So I would like to understand that better, and then adjust the code accordingly. But this can be done separately.

Reply:
The catchVoidToResult is only around the release.
If we want to we could move the self->getStream() outside.
But we do not need this anywhere after this catch so I do not see the point